cpp/src/arrow/io/file.cc

```diff
@@ -408,14 +429,13 @@ class OSFile {
   int fd() const { return fd_; }
 
   bool is_open() const { return is_open_; }
-  const std::string& path() const { return path_; }
 
   int64_t size() const { return size_; }
 
   FileMode::type mode() const { return mode_; }
 
  protected:
-  std::string path_;
+  std::unique_ptr<PlatformFilename> path_;
```

Comment:
I wanted it to be immutable for no practical reason, since this is a private class. Refactored things to use not unique_ptr